Convert lock unlock to generics

manzoorwanijk opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Addresses the feedback given in https://github.com/WordPress/gutenberg/pull/66667#discussion_r1825958215

## Why?
To allow better type support for `lock` and `unlock` functions

## How?
This PR makes `lock` and `unlock` as generics to set the object type

## Testing Instructions
- Same as #66667

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

manzoorwanijk committed:
Convert lock unlock to generics

manzoorwanijk committed:
Set object type from generic for unlock

manzoorwanijk commented:
@jsnajdr Should I fix those types that are affected by the updated signature?

There are 7 errors in total, fixable by 3-4 minor changes

manzoorwanijk committed:
Fix types affected by updated signature of lock/unlock

manzoorwanijk commented on packages/components/src/private-apis.ts:
```diff
@@ -10,13 +10,16 @@ import { Tabs } from './tabs';
-export const privateApis = {};
-lock( privateApis, {
+const privateData = {
 	__experimentalPopoverLegacyPositionToPlacement,
 	createPrivateSlotFill,
 	ComponentsContext,
 	Tabs,
 	Theme,
 	Menu,
 	kebabCase,
-} );
+};
+
+export const privateApis = {} as typeof privateData;
+
+lock( privateApis, privateData );
```
This is only to set the proper type for `privateApis` here to ensure the consumer components that use `lock( privateApis )` receive proper types.

manzoorwanijk committed:
Merge remote-tracking branch 'upstream/trunk' into update/convert-lock-unlock-to-generics

youknowriad commented on packages/private-apis/src/implementation.ts:
```diff
@@ -170,17 +172,19 @@ function lock( object: Record< symbol, WeakKey >, privateData: unknown ) {
  * @return The private data bound to the object.
  */
-function unlock( object: Record< symbol, WeakKey > ) {
+function unlock< T >( object: T ): T {
```
This looks wrong to me personally, unlock doesn't return the same "type" of the object you give it.

manzoorwanijk replied:
It was as per the suggestion by @jsnajdr in https://github.com/WordPress/gutenberg/pull/66667#discussion_r1825958215

I think the logic inside `lock` and `unlock` should be alien to the consumer code and it should/need not know what it internally does, because we expect them to use `lock` `unlock` together, not in isolation.

youknowriad replied:
But this is different from @jsnajdr's suggestion, the type of the argument and the returned value don't have the same types there.

manzoorwanijk replied:
Then, in that case, there will be no inference from TypeScript, and the consumer will always have to pass the type as the generic to ensure that the return type is not always `unknown,` which can be annoying.

youknowriad replied:
I know but if we return the same type, it's wrong too because we'll either have locked items missing from the returned value if the returned object contains the original one (which happens for store), or we'll have a completely wrong return type.

I see things closer to this but I'm not sure how doable:

```
 lock( o: X, privateApis:  Y ): XwithLockedY
 unlock( XwithLockedY ):  X & Y
```

manzoorwanijk replied:
Thanks, I have updated the signature to `unlock< T = any >( object: unknown ): T` to use `any` as default to avoid breaking existing types.

youknowriad replied:
Thanks, I'll let others, more experts with TS chime in. cc @jsnajdr @sirreal if needed.

manzoorwanijk committed:
Improve signature

manzoorwanijk committed:
Remove expected errors no longer needed

manzoorwanijk committed:
Restore the type for component private APIs

manzoorwanijk commented:
@jsnajdr this one addresses your feedback given in https://github.com/WordPress/gutenberg/pull/66667#discussion_r1825958215.

Can we get this through?

youknowriad reviewed (approved):
Works for me. I'd appreciate @jsnajdr's thoughts too.

jsnajdr reviewed (approved):
Yes this looks good. We don't have any real-world usages of the types yet, so we are not sure that they are the correct ones, but we can patch that when some problem is uncovered.

manzoorwanijk commented:
Thank you folks. I can't merge PRs here. Please go ahead whenever you want.

tyxla reviewed (approved):
Thanks @manzoorwanijk, this looks great! I'll get it shipped 🚀
